### PR TITLE
padding and cropping classes use MetaTensor

### PR DESCRIPTION
**Pull request by rijobro**
### Description
Pad transforms use MetaTensor to allow inverse.

also fixes https://github.com/Project-MONAI/MONAI/issues/4435

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.

### TODO

Name | Array implementation | Dict implementation | Array test | Dict test
:------------ | :-------------| :-------------| :-------------| :-------------
SpatialPad | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
BorderPad | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
DivisiblePad | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
SpatialCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
CenterSpatialCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
CenterScaleCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandSpatialCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandScaleCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandSpatialCropSamples | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
CropForeground | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandWeightedCrop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandCropByPosNegLabel | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
RandCropByLabelClasses |  :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:
ResizeWithPadOrCop | :heavy_check_mark: |  :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:

### CODE REVIEW

File: monai/transforms/croppad/array.py [hunk 1]
```diff
@@ -109,8 +107,87 @@ def _pt_pad(img: torch.Tensor, all_pad_width, mode, **kwargs) -> torch.Tensor:
...
+        # but if mode doesn't exist in pytorch, use numpy
+        except (ValueError, TypeError) as err:
+            if "Unsupported option" in str(err) or "unexpected keyword" in str(err):
+                # extract metadata
+                img_np = img.detach().cpu().numpy()
+                mode = convert_pad_mode(dst=img_np, mode=mode or self.mode).value
+                out = torch.as_tensor(self._np_pad(img_np, to_pad, mode, **self.kwargs))
+                if get_track_meta():
+                    out = MetaTensor(out, meta=img.meta, applied_operations=img.applied_operations)  # type: ignore
+        if isinstance(img, MetaTensor):
```
[the remark this answers is not in the document]
Comment (wyli) [reply monai/transforms/croppad/array.py:156]: cropping looks good...
```py
from monai.transforms import CenterSpatialCrop, EnsureChannelFirst, LoadImage, SaveImage

filename = "~/Downloads/avg152T1_LR_nifti.nii.gz"
img = LoadImage()(filename)
img = EnsureChannelFirst()(img)
img = CenterSpatialCrop(roi_size=(40, 50, 60))(img)
SaveImage(resample=False)(img)
```

Screenshot 2022-05-31 at 14 20 33

[the remark this answers is not in the document]
Comment (rijobro) [reply monai/transforms/croppad/array.py:156]: @wyli this is great, thanks for this!

**Comment by rijobro** [comment]
@wyli PR is getting very big as I'm doing all padding and cropping. Apologies for that, it's a hard one to split.

File: tests/test_pad.py [hunk 2]
```diff
@@ -111,23 +111,25 @@ def get_info(im: MetaTensor):
         def check(info1, info2, should_be_same):
             aff1, meta1, app_ops1 = info1
             aff2, meta2, app_ops2 = info2
-            # meta and app_ops always same
-            self.assertEqual(meta1, meta2)
```
Comment (wyli) [inline tests/test_pad.py]: Hi @rijobro I put some tests here (including the inverse). `tests/test_pad.py` and `tests/tests.test_crop_base.py` should work now.  (this `self.assertEqual(meta1, meta2)` is not guaranteed because there is info about applied operations.)

**Comment by rijobro** [comment]
@wyli thanks for the help here. I'm off for a week, feel free to continue this without me if you want to. All transforms have been updated, just need to get all the linting happy. I suspect https://github.com/Project-MONAI/MONAI/issues/4435 will need to be addressed first.

**Comment by wyli** [comment]
> @wyli thanks for the help here. I'm off for a week, feel free to continue this without me if you want to. All transforms have been updated, just need to get all the linting happy. I suspect #4435 will need to be addressed first.

thanks! I'll try to merge this one soon.

**Comment by wyli** [comment]
cropping with multiple random samples (basic `applied_operations` collating/decollating work fine)
```py
from monai.transforms import EnsureChannelFirst, LoadImage, RandSpatialCropSamples, SaveImage

filename = "~/Downloads/avg152T1_LR_nifti.nii.gz"
img = LoadImage()(filename)
img = EnsureChannelFirst()(img)
img = RandSpatialCropSamples(roi_size=(40, 50, 60), random_size=False, num_samples=3)(img)
for i in img:
    SaveImage(resample=False)(i)

```
Screenshot 2022-06-06 at 16 23 58

**Review by wyli — APPROVED** [review]
merging this for now as the basic functionality works nicely, but it will need another discussion/refactoring for the design of modules such as the `PadBase` class https://github.com/Project-MONAI/MONAI/blob/6c56810cdc8d8f7a638daebebe7d4486f296f26f/monai/transforms/croppad/array.py#L80 cc @Nic-Ma @ericspod
